solr/core/src/java/org/apache/solr/storage/DirectBufferPool.java

```diff
@@ -19,46 +19,105 @@
+  private final LongAdder discarded = new LongAdder();
+  private final AtomicLong poolSize = new AtomicLong();
+  private final LongAdder hits = new LongAdder();
+  private final AtomicInteger outstanding = new AtomicInteger();
+  private final AtomicInteger outstandingHighWatermark = new AtomicInteger();
+
+  private final int maxPoolSize;
+
+  public DirectBufferPool(int size, int blockSize, int allocateChunkSize) {
+    this.maxPoolSize =
+        Math.min(
+            8192, Math.toIntExact((Runtime.getRuntime().maxMemory() / (size + blockSize - 1)) / 4));
     this.size = size;
     this.blockSize = blockSize;
-    this.allocateChunkSize = allocatChunkSize;
+    this.allocateChunkSize = allocateChunkSize;
     if (size % blockSize != 0) {
       throw new IllegalArgumentException("block size incompatible with size");
     }
   }
 
+  long[] getStats() {
+    return new long[] {
+      created.sum(),
+      poolSize.get(),
+      hits.sum(),
+      discarded.sum(),
+      outstanding.get(),
+      outstandingHighWatermark.get()
+    };
+  }
+
+  private void incrementOutstanding() {
+    int newSize = outstanding.incrementAndGet();
+    int extantHighWatermark = outstandingHighWatermark.get();
+    if (newSize > extantHighWatermark) {
+      do {
+        extantHighWatermark =
+            outstandingHighWatermark.compareAndExchange(extantHighWatermark, newSize);
+      } while (extantHighWatermark < newSize);
+    }
+  }
+
   public ByteBuffer get() {
     ByteBuffer ret = pool.poll();
     if (ret != null) {
+      poolSize.decrementAndGet();
+      hits.increment();
+      incrementOutstanding();
       return ret;
     } else {
       synchronized (pool) {
         ret = pool.poll();
         if (ret != null) {
+          poolSize.decrementAndGet();
+          hits.increment();
+          incrementOutstanding();
           return ret;
         }
         ByteBuffer chunk =
             ByteBuffer.allocateDirect((size * allocateChunkSize) + blockSize - 1)
                 .alignedSlice(blockSize);
         int pos = 0;
-        for (int i = 1; i < allocateChunkSize; i++) {
-          pool.add(chunk.position(pos).limit(pos += size).slice());
+        int added = 0;
+        try {
+          for (int i = 1; i < allocateChunkSize; i++) {
+            pool.add(chunk.position(pos).limit(pos += size).slice());
+            poolSize.incrementAndGet();
+            added++;
+          }
+        } finally {
+          created.add(added);
         }
-        return chunk.position(pos).limit(pos + size).slice();
+        ret = chunk.position(pos).limit(pos + size).slice();
+        created.increment();
+        incrementOutstanding();
+        return ret;
       }
     }
   }
 
   public void release(ByteBuffer bb) {
+    outstanding.decrementAndGet();
     bb.clear();
-    pool.offer(bb);
+    if (poolSize.incrementAndGet() > maxPoolSize) {
+      poolSize.decrementAndGet();
+      discarded.increment();
+    } else {
+      pool.add(bb);
```

Comment:
Should this remain as offer to avoid an exception? What is the harm if offer were to fail? I know we check above, but with multiple threads another could add between our check and add, right?

Reply:
Good callout; ConcurrentLinkedQueue is guaranteed to always return true/not throw an exception (as it says in the overridden javadocs for `add()` on that class -- there's no capacity restriction). The queue may briefly exceed the limit we have in place, but it's a benign race condition.

Reply:
Ah gotcha, was looking at the docs for just a queue, that makes sense.